### ISSUE TITLE
Hyprland crashes when the monitor is turned off.

### ISSUE THREAD

**Issue by NotAShelf** [hyprwm/Hyprland#2646]
Hyprland crashes whenever I turn off my monitor, or from what I understand it crashes exactly when it's powered back on.

The monitor is properly configured in config, plugged via HDMI. Other HDMI monitor works fine.

[Crash Log](https://p.frzn.dev/39q)

```bash
$ hyprctl monitors
Monitor HDMI-A-1 (ID 0):
	1920x1080@60.000000 at 0x0
	description: ASUSTek COMPUTER INC ASUS VP249 R1LMTF049118 (HDMI-A-1)
	make: ASUSTek COMPUTER INC
	model: ASUS VP249
	serial: R1LMTF049118
	active workspace: 1 (1)
	special workspace: 0 ()
	reserved: 64 0 0 0
	scale: 1.00
	transform: 0
	focused: yes
	dpmsStatus: 1
	vrr: 0
```

According to the crashlog, the monitor is removed once it's powered off.

```bash
[LOG] Monitor HDMI-A-1 layers arranged: reserved: 0.000000 0.000000 0.000000 0.000000
[LOG] Removed monitor HDMI-A-1!
```

**Comment by mabequinho** [comment]
Same issue.

```
Monitor HDMI-A-1 (ID 0):
	2560x1080@74.987000 at 0x0
	description: LG Electronics LG ULTRAWIDE 0x01010101 (HDMI-A-1)
	make: LG Electronics
	model: LG ULTRAWIDE
	serial: 0x01010101
	active workspace: 1 (1)
	special workspace: 0 ()
	reserved: 0 0 0 32
	scale: 1.00
	transform: 0
	focused: yes
	dpmsStatus: 1
	vrr: 0
```

`VGA compatible controller: Advanced Micro Devices, Inc. [AMD/ATI] Lexa PRO [Radeon 540/540X/550/550X / RX 540X/550/550X] (rev c7)`

**Comment by mathologist** [comment]
I've been having the same issue. I'm on Arch using hyprland-git from the AUR, but I get the same problem when I use the release version. I've attached my crash report.

[hyprlandCrashReport121118.txt](https://github.com/hyprwm/Hyprland/files/11973769/hyprlandCrashReport121118.txt)

**Comment by vaxerski** [comment]
bisect? is this a regression?

@NotAShelf garbage crash report (nix moment)

@mathologist wlroots issue perhaps? doesn't seem  to have any hyprland frames. Might be memory faults as well.

**Comment by NotAShelf** [comment]
> bisect? is this a regression?

Regression, don't think so. It works for a monitor and doesn't for another

**Comment by fettywaptrapqueen** [comment]
Can reproduce on -git, let me know if there's anything I can do to help with debugging (warning: I'm retar)ed

**Comment by Glitch1441** [comment]
Can say that the same happens here.

**Comment by romanstingler** [comment]
guys who installed -git package 
can you check your real version 
`hyprctl version` 
At least for me the git package seems not to upgrade the package properly.

**Comment by jessesung** [comment]
Confirmed that after rolling back to 0.26.0-3 the issue can't be reproduced any more.

If you have a monitor connected via DP, besides switching on/off the monitor, you can also reproduce the issue by
```
hyprctl dispatch dpms off ; sleep 10 ; hyprctl dispatch dpms on
```

**Comment by vaxerski** [comment]
`git bisect` then pleas

**Comment by jessesung** [comment]
@vaxerski: `git bisect` shows:
```
01f85a09a981eb6c8784d459f729d418c0dd19e3 is the first bad commit
commit 01f85a09a981eb6c8784d459f729d418c0dd19e3
Author: vaxerski <vaxry@vaxry.net>
Date:   Fri Jun 23 21:14:04 2023 +0200

    xwayland: send zero scaling to xwayland if enabled
```

Also after reverting this commit from v0.27.0, the issue can no longer be reproduced.

**Comment by SkyLeite** [comment]
To add another data point, I'm suffering from the same issue. When I `hyprctl dpms off` and go to bed, 10/10 times when I wake my computer up in the morning I'm back at my login manager. For some reason the backtrace in my crash report has less information than the ones posted in the thread however.

[hyprlandCrashReport2635.txt](https://github.com/hyprwm/Hyprland/files/12062051/hyprlandCrashReport2635.txt)

**Comment by vaxerski** [comment]
[patch.txt](https://github.com/hyprwm/Hyprland/files/12062616/patch.txt)
try this patch

**Comment by jessesung** [comment]
> [patch.txt](https://github.com/hyprwm/Hyprland/files/12062616/patch.txt) try this patch

Nope, doesn't work...

**Comment by SkyLeite** [comment]
> [patch.txt](https://github.com/hyprwm/Hyprland/files/12062616/patch.txt) try this patch

Seems to work for me! Just had to edit the patch to remove the (seemingly) unnecessary diff on the `wlroots` submodule

**Comment by vaxerski** [comment]
> (seemingly) unnecessary diff on the wlroots submodule

git does that

**Comment by SkyLeite** [comment]
Yeah, I figured. Still had to remove it for it to work on NixOS, since the wlroots module doesn't exist then, but it's not a big deal. Just thought I'd mention it

**Comment by vaxerski** [comment]
pushed in 3229862dd4cbfa93638a4d16ed86ec2fda5d38a6, lmk

**Comment by mathologist** [comment]
Unfortunately, this didn't fix it for me. I compiled it myself from git and ran it with the default config in case it's something else. I'm wondering if I have a different problem.  I don't know if it's a memory issue since it's completely consistent and did only start for me since upgrading to 0.27. 

I've attached my log and crash report, just in case it helps.
[hyprlandCrashReport281155.txt](https://github.com/hyprwm/Hyprland/files/12063005/hyprlandCrashReport281155.txt)
[hyprland.log](https://github.com/hyprwm/Hyprland/files/12063006/hyprland.log)

**Comment by SkyLeite** [comment]
Seems like it was just a red herring for me. I hit the issue again, with the exact same crash report as @mathologist's

**Comment by jessesung** [comment]
Hi @vaxerski,

Both m->output and m->m_bEnable are valid when the issue happens hence the patch doesn't prevent the issue from happening. Since this seems to be happened when the monitor is back, it feels reasonable.

Though I also found this in the log:
```
[LOG] Pending output DP-1 does not accept VRR.
```
while
```
> hyprctl monitors | grep vrr
	vrr: 1
```
Wondering if there's a race between CHyprXWaylandManager::setXWaylandScale and something else.

**Comment by jessesung** [comment]
Both `CHyprRenderer::applyMonitorRule()` and `Events::listener_monitorCommit()` call `g_pXWaylandManager->updateXWaylandScale()`, that's why `setXWaylandScale()` is being called twice in the crash report.

**Comment by vaxerski** [comment]
can someone get that crash report from a debug build?

**Comment by jessesung** [comment]
> can someone get that crash report from a debug build?

Not sure if I did it correctly, Hyprland did start with a message saying that it's using a default config `.hyprlandd.conf` though.

```
Backtrace:
        #0 | Hyprland(_ZN13CrashReporter18createAndSaveCrashEi+0xa68) [0x55a941a7f3a8]
                ??
                ??:0
        #1 | Hyprland(_Z25handleUnrecoverableSignali+0x44) [0x55a941a07a24]
                ??
                ??:0
        #2 | /usr/lib/libc.so.6(+0x39ab0) [0x7f6b3ba4fab0]
                ??
                ??:0


Log tail:
[LOG] Searching for matching rules for kitty (title: ~)
[LOG] Window 55a94433b010 set title to ~
[LOG] Destroy called for monitor DP-1
[LOG] onDisconnect called for DP-1
[LOG] Callback 55a943d9c7d0 -> 55a943d9c7c8,  removed.
[LOG] Callback 55a943d9c908 -> 55a943d9c900,  removed.
[LOG] Callback 55a943d9c970 -> 55a943d9c968,  removed.
[LOG] Callback 55a943d9c9d8 -> 55a943d9c9d0,  removed.
[LOG] Removed monitor DP-1!
[LOG]  [hookSystem] New hook event registered: monitorRemoved
[WARN] Unplugged last monitor, entering an unsafe state. Good luck my friend.
[LOG] Callback 55a943d9c8a0 -> 55a943d9c898,  removed.
[LOG] Callback 55a943d9c838 -> 55a943d9c830,  removed.
[WARN] Recovering from an unsafe state. May you be lucky.
[LOG] Recovering a removed monitor.
[LOG] Callback 55a943d9ca40 -> 55a943d9ca38,  removed.
[LOG] Registered signal for owner 55a943d9c590: 55a943d8b770 -> 55a943d9c7a8 (owner: )
[LOG] Registered signal for owner 55a943d9c590: 55a943d8b800 -> 55a943d9c810 (owner: )
[LOG] Registered signal for owner 55a943d9c590: 55a943d8b7f0 -> 55a943d9c878 (owner: )
[LOG] Registered signal for owner 55a943d9c590: 55a943d8b780 -> 55a943d9c8e0 (owner: )
[LOG] Registered signal for owner 55a943d9c590: 55a943d8b790 -> 55a943d9c948 (owner: )
[LOG] Registered signal for owner 55a943d9c590: 55a943d8b7b0 -> 55a943d9c9b0 (owner: )
[LOG] Registered signal for owner 55a943d9c590: 55a943d8b7d0 -> 55a943d9ca18 (owner: )
[WARN] No rule found for DP-1, trying to use the first.
[LOG] Applying monitor rule for DP-1
[LOG] Setting preferred mode for DP-1
[LOG] Overriding XWayland scale with 0.00
[LOG] Monitor DP-1 -> destroyed all render data
[LOG] Monitor DP-1 layers arranged: reserved: 0.000000 0.000000 0.000000 0.000000
[LOG] Monitor DP-1 data dump: res 3840x2160@120.00Hz, scale 1.00, transform 0, pos 0x0, 10b 0
[LOG] Overriding XWayland scale with 0.00
[LOG] Added new monitor with name DP-1 at 0,0 with size 3840x2160, pointer 55a943d8b620
[LOG] New monitor: WORKSPACEID 2, exists: 0
[LOG] Monitor DP-1 layers arranged: reserved: 0.000000 0.000000 0.000000 0.000000
[WARN] No rule found for DP-1, trying to use the first.
[LOG] Applying monitor rule for DP-1
[LOG] Setting preferred mode for DP-1
[LOG] Monitor DP-1 -> destroyed all render data
[LOG] Monitor DP-1 layers arranged: reserved: 0.000000 0.000000 0.000000 0.000000
[LOG] Monitor DP-1 data dump: res 3840x2160@120.00Hz, scale 1.00, transform 0, pos 0x0, 10b 0
[LOG] Overriding XWayland scale with 0.00
[LOG] Monitor DP-1 layers arranged: reserved: 0.000000 0.000000 0.000000 0.000000
[LOG] Framebuffer created, status 36053
[LOG] Framebuffer created, status 36053
[LOG] Framebuffer created, status 36053
[LOG] Framebuffer created, status 36053
[LOG] Allocated texture for BGTex
[LOG] Background created for monitor DP-1
[LOG] Shaders initialized successfully.
[LOG] Overriding XWayland scale with 0.00
```

**Comment by vaxerski** [comment]
doesn't help that the backtrace is fucked. What does `coredumpctl debug <pid>` and then `bt -full` say?

**Comment by jessesung** [comment]
```
(gdb) bt -full
#0  0x00007f65d63b226c in  () at /usr/lib/libc.so.6
#1  0x00007f65d6362a08 in raise () at /usr/lib/libc.so.6
#2  0x00007f65d634b538 in abort () at /usr/lib/libc.so.6
#3  0x00005646fa7e8a2a in  ()
#4  0x00007f65d6362ab0 in <signal handler called> () at /usr/lib/libc.so.6
#5  0x0000000000000000 in  ()
#6  0x00007f65d6cc101e in wl_signal_emit_mutable () at /usr/lib/libwayland-server.so.0
#7  0x00007f65d6dd61d0 in  () at /usr/lib/libwlroots.so.12032
#8  0x00007f65d5ea94f6 in  () at /usr/lib/libffi.so.8
#9  0x00007f65d5ea5f5e in  () at /usr/lib/libffi.so.8
#10 0x00007f65d5ea8b73 in ffi_call () at /usr/lib/libffi.so.8
#11 0x00007f65d6cbfada in  () at /usr/lib/libwayland-server.so.0
#12 0x00007f65d6cc4180 in  () at /usr/lib/libwayland-server.so.0
#13 0x00007f65d6cc2ae2 in wl_event_loop_dispatch () at /usr/lib/libwayland-server.so.0
#14 0x00007f65d6cc32d7 in wl_display_run () at /usr/lib/libwayland-server.so.0
#15 0x00005646fa8c835c in main ()
```

**Comment by vaxerski** [comment]
memory corruption, hm

**Comment by vaxerski** [comment]
can someone build with ASan and reproduce? Looks like some UAF or shit

```
cmake --no-warn-unused-cli -DCMAKE_BUILD_TYPE:STRING=Debug -H./ -B./build -G Ninja -DWITH_ASAN:STRING=True
cmake --build ./build --config Debug --target all -j`nproc 2>/dev/null || getconf NPROCESSORS_CONF`

# launch hyprland
ASAN_OPTIONS="log_path=asan.log" ./build/Hyprland
```

once it crashes, look for a file called `asan.log.xxxxx` in either cwd or `~`

**Comment by jessesung** [comment]
Although the build seems to be correct:
```
> rg -i asan build
build/build.ninja
95:  LINK_LIBRARIES = <...> -lasan <...>
```
I'm unable to find any ASan log after it crashes...

**Comment by vaxerski** [comment]
did you set the envvar? Launch from the tty as well, maybe it doesn't propagate. If it won't, the logs will be printed to stderr.

you can do a `find` in your `~`, it sometimes somehow goes to weird places.

**Comment by jessesung** [comment]
I ran `env ASAN_OPTIONS="log_path=asan.log" ./build/Hyprland` in a tty.

**Comment by vaxerski** [comment]
odd. try removing the env and watching stderr then.

**Comment by Cloudef** [comment]
```
* thread #1, name = '.Hyprland-wrapp', stop reason = signal SIGABRT
  * frame #0: 0x00007fc478c80a8c libc.so.6`__pthread_kill_implementation + 268
    frame #1: 0x00007fc478c31c86 libc.so.6`raise + 22
    frame #2: 0x00007fc478c1b8ba libc.so.6`abort + 215
    frame #3: 0x000000000064bef1 .Hyprland-wrapped`handleUnrecoverableSignal(int) + 110
    frame #4: 0x00007fc478c31d30 libc.so.6`__restore_rt
    frame #5: 0x00000000031d0d80
    frame #6: 0x00007fc479848a1c libwayland-server.so.0`wl_signal_emit_mutable + 124
    frame #7: 0x00007fc4797915f3 libwlroots.so.12`output_bind + 227
    frame #8: 0x00007fc478bf3a3a libffi.so.8`ffi_call_unix64 + 82
    frame #9: 0x00007fc478bf2935 libffi.so.8`ffi_call_int + 501
    frame #10: 0x00007fc478bf3518 libffi.so.8`ffi_call + 216
    frame #11: 0x00007fc47984c886 libwayland-server.so.0`wl_closure_invoke + 390
    frame #12: 0x00007fc479847c4b libwayland-server.so.0`wl_client_connection_data + 587
    frame #13: 0x00007fc47984a8f2 libwayland-server.so.0`wl_event_loop_dispatch + 162
    frame #14: 0x00007fc479848455 libwayland-server.so.0`wl_display_run + 37
    frame #15: 0x0000000000650024 .Hyprland-wrapped`CCompositor::startCompositor() + 1012
    frame #16: 0x000000000085f436 .Hyprland-wrapped`main + 1811
    frame #17: 0x00007fc478c1cace libc.so.6`__libc_start_call_main + 126
    frame #18: 0x00007fc478c1cb89 libc.so.6`__libc_start_main@@GLIBC_2.34 + 137
    frame #19: 0x000000000064bd65 .Hyprland-wrapped`_start + 37
 ```
Stacktrace with LLDB
Can be reproduced when used in combination with way-displays
https://github.com/Cloudef/nixos-flake/blob/master/modules/hyprland-desktop.nix

**Comment by vaxerski** [comment]
work on this in #2733

**Comment by Cloudef** [comment]
```
(lldb) bt
* thread #1, name = '.Hyprland-wrapp', stop reason = signal SIGABRT
  * frame #0: 0x00007f46b51a0a8c libc.so.6`__pthread_kill_implementation + 268
    frame #1: 0x00007f46b5151c86 libc.so.6`raise + 22
    frame #2: 0x00007f46b513b8ba libc.so.6`abort + 215
    frame #3: 0x0000000000474f54 .Hyprland-wrapped`handleUnrecoverableSignal(int) + 68
    frame #4: 0x00007f46b5151d30 libc.so.6`__restore_rt
    frame #5: 0x0000000000000000
    frame #6: 0x00007f46b5d79a1c libwayland-server.so.0`wl_signal_emit_mutable + 124
    frame #7: 0x00007f46b5cc25f3 libwlroots.so.12`output_bind + 227
    frame #8: 0x00007f46b5633a3a libffi.so.8`ffi_call_unix64 + 82
    frame #9: 0x00007f46b5632935 libffi.so.8`ffi_call_int + 501
    frame #10: 0x00007f46b5633518 libffi.so.8`ffi_call + 216
    frame #11: 0x00007f46b5d7d886 libwayland-server.so.0`wl_closure_invoke + 390
    frame #12: 0x00007f46b5d78c4b libwayland-server.so.0`wl_client_connection_data + 587
    frame #13: 0x00007f46b5d7b8f2 libwayland-server.so.0`wl_event_loop_dispatch + 162
    frame #14: 0x00007f46b5d79455 libwayland-server.so.0`wl_display_run + 37
    frame #15: 0x000000000047421c .Hyprland-wrapped`main + 2460
    frame #16: 0x00007f46b513cace libc.so.6`__libc_start_call_main + 126
    frame #17: 0x00007f46b513cb89 libc.so.6`__libc_start_main@@GLIBC_2.34 + 137
    frame #18: 0x0000000000474e05 .Hyprland-wrapped`_start + 37
    ```
    seems to still crash similarly

**Comment by Cloudef** [comment]
With the correct rev, does not change much
```gdb
#0  0x00007f98654a1a8c in __pthread_kill_implementation () from /nix/store/ayg065nw0xi1zsyi8glfh5pn4sfqd8xg-glibc-2.37-8/lib/libc.so.6
No symbol table info available.
#1  0x00007f9865452c86 in raise () from /nix/store/ayg065nw0xi1zsyi8glfh5pn4sfqd8xg-glibc-2.37-8/lib/libc.so.6
No symbol table info available.
#2  0x00007f986543c8ba in abort () from /nix/store/ayg065nw0xi1zsyi8glfh5pn4sfqd8xg-glibc-2.37-8/lib/libc.so.6
No symbol table info available.
#3  0x00000000004ac12d in handleUnrecoverableSignal (sig=11) at ../src/Compositor.cpp:35
No locals.
#4  <signal handler called>
No symbol table info available.
#5  0x00007f98661864c4 in wl_list_insert () from /nix/store/3y9pj04f7xfwfs6h9rvgd70cqi105g54-wayland-1.22.0/lib/libwayland-server.so.0
No symbol table info available.
#6  0x00007f9866181a13 in wl_signal_emit_mutable () from /nix/store/3y9pj04f7xfwfs6h9rvgd70cqi105g54-wayland-1.22.0/lib/libwayland-server.so.0
No symbol table info available.
#7  0x00007f98660a56b1 in output_bind (wl_client=0x2158600, data=0x24818c0, version=1, id=5) at ../types/output/output.c:128
        output = 0x24818c0
        resource = 0x21d9890
        evt = {output = 0x24818c0, resource = 0x21d9890}
#8  0x00007f98658fda3a in ffi_call_unix64 () from /nix/store/v4gixcxxx0dwy8s8l32i63cbnz16zn5d-libffi-3.4.4/lib/libffi.so.8
No symbol table info available.
#9  0x00007f98658fc935 in ffi_call_int () from /nix/store/v4gixcxxx0dwy8s8l32i63cbnz16zn5d-libffi-3.4.4/lib/libffi.so.8
No symbol table info available.
#10 0x00007f98658fd518 in ffi_call () from /nix/store/v4gixcxxx0dwy8s8l32i63cbnz16zn5d-libffi-3.4.4/lib/libffi.so.8
No symbol table info available.
#11 0x00007f9866185886 in wl_closure_invoke () from /nix/store/3y9pj04f7xfwfs6h9rvgd70cqi105g54-wayland-1.22.0/lib/libwayland-server.so.0
No symbol table info available.
#12 0x00007f9866180c4b in wl_client_connection_data () from /nix/store/3y9pj04f7xfwfs6h9rvgd70cqi105g54-wayland-1.22.0/lib/libwayland-server.so.0
No symbol table info available.
#13 0x00007f98661838f2 in wl_event_loop_dispatch () from /nix/store/3y9pj04f7xfwfs6h9rvgd70cqi105g54-wayland-1.22.0/lib/libwayland-server.so.0
No symbol table info available.
#14 0x00007f9866181455 in wl_display_run () from /nix/store/3y9pj04f7xfwfs6h9rvgd70cqi105g54-wayland-1.22.0/lib/libwayland-server.so.0
No symbol table info available.
#15 0x00000000004b0411 in CCompositor::startCompositor (this=0x17fa650) at ../src/Compositor.cpp:479
No locals.
#16 0x00000000005a8bcb in main (argc=<optimized out>, argv=<optimized out>) at ../src/main.cpp:96
        cmd = "/nix/store/dw4s4y799gnqj7zc24zc55s7zl6bjy15-hyprland-debug-0.27.0+date=2023-07-17_24327b7/bin/Hyprland --config /etc/xdg/hyprland.conf"
        configPath = "/etc/xdg/hyprland.conf"
```

**Comment by Cloudef** [comment]
Attached is WAYLAND_DEBUG=1 output
[hyprland.log](https://github.com/hyprwm/Hyprland/files/12075202/hyprland.log)

**Comment by vaxerski** [comment]
we have already estabilished it's a memory issue, saying that again won't help much. You should move to #2733 and do tests there.

**Comment by vaxerski** [comment]
brother b156a965 is 2 days old

**Comment by jessesung** [comment]
Interesting... My main repo is up-to-date while the clone I built against is not... Let me try again.

**Comment by iynaix** [comment]
The issue appears to be resolved for me, no longer crashes when I turn my monitors off and back on.

**Comment by jessesung** [comment]
Made a clean build with `16fd9084` and confirmed the issue is fixed. Thanks!!

**Comment by Cloudef** [comment]
Confirming here as well. Now if it was possible for the clients to survive KVM switch as well, then it would be perfect, should I open new issue for this?

**Comment by trougnouf** [comment]
Building master fixes the crash back to TTY for me but then I get back a desktop with the default wallpaper, no waybar, and no response (except for the Super+m to quit).

**Comment by istasi** [comment]
https://pastebin.com/D4V1nERY seemingly same issue, whenver monitor is off in anyway, pc suspend, monitor standby or straight turned off. All i get is hyprchan, none of the previous opened programs, seemingly only default config.

**Comment by NotAShelf** [comment]
please stop necrobumping this issue. if it's still a problem, open a new issue after making sure you are on the latest possible version.
